src/main/webapp/script.js

```diff
@@ -0,0 +1,382 @@
+    rectsToBlurPromise.then((rects) => {
+      const blurredImage = getBlurredImage(rects, imageObj);
+      createDownloadButton(blurredImage);
+      showBlurredImage(blurredImage);
+    });
+  };
+}
+
+/**
+ * Fucntion to update html page for blurring a new image.
+ * It resizes canvases and draws new image on them.
+ * It deletes previous download button if there was.
+ * @param {Image} image
+ */
+function prepareHtmlElements(image) {
+  const prepareCanvas = (canvasId) => {
+    const canvas = document.getElementById(canvasId);
+    const ctx = canvas.getContext('2d');
+
+    // clear canvas
+    ctx.clearRect(0, 0, canvas.width, canvas.height);
+
+    // resize canvas height to fit new image
+    canvas.height = image.height * canvas.width / image.width;
+
+    // draw new image on it, scaling the image to fit in canvas
+    ctx.drawImage(image, 0, 0, canvas.width, canvas.height);
+  };
+  prepareCanvas('input-canvas');
+  prepareCanvas('output-canvas');
+
+  // remove download button
+  const downloadButton = document.getElementById('download-button');
+  if (downloadButton !== null) {
+    downloadButton.remove();
+  }
+}
+
+/**
+ * Class for a rectangle to blur
+ */
+class Rect {
+  /**
+   * @param {function} func Can be min or max function
+   * @param {string} property Property which all the items
+   *                          in arrayObject have
+   * @param {Array} arrayObject array of Objects
+   * @return {Number} min or max (depending on func) value of the
+   *         specified property among arrayObject items.
+   */
+  static get(func, property, arrayObject) {
+    let ans = arrayObject[0][property];
+
+    for (const item of arrayObject) {
+      ans = func(ans, item[property]);
+    }
+
+    return ans;
+  }
+
+  /**
+   * @constructor
+   * @param {JsonRect} _rect
+   * Constructs Rect from rectangle object we get from json
+   * Rectange object we get from json has 4 points - its corners,
+   * our Rect has top left corner, width and height.
+   */
+  constructor(_rect) {
+    this.leftX = Rect.get(Math.min, 'x', _rect);
+    this.topY = Rect.get(Math.min, 'y', _rect);
+    this.width = Rect.get(Math.max, 'x', _rect) - this.leftX;
+    this.height = Rect.get(Math.max, 'y', _rect) - this.topY;
+  }
+}
+
+/**
+ * Function that blurs specific rectangles of an image
+ * @param {Array<JsonRect>} rectsToBlur A list of coordinates of
+ *                                      rectangles to blur
+ * @param {Image} image Image on which we want to blur those rectangles
+ * @return {HTMLCanvasElement} canvas with an image with blurred rectangles
+ * The general idea is to create canvas with blurred image on it,
+ * and canvas with not blurred image on it. Then we need to extract
+ * rectangles from blurred image and put them on top of not-blurred image.
+ * The image is blurred using Gaussian blur algorithm
+ * (it is implemented in canvas.filter.blur)
+ */
+function getBlurredImage(rectsToBlur, image) {
```

Comment:
Another approach I would suggest is to create a smaller canvas for every rectangle you want to blur (containing only the rectangle) and getting the pixels from there instead of blurring the whole image.

Reply:
I am not sure if this will be faster, because in this approach I will need to use getImageData and putImageData methods instead of drawImage in order to draw image on blurred canvas. It seems like your approach will be faster, though, but I'd like to measure it and there can also be some other optimisations to think of.
Now blurring doesn't take much time, so I think I will make another pull request with experiments on the speed and memory of this function later, because it takes some research to do.
Thank you for the idea! If you think of any other optimisations to include in that pull request, please let me know )

Reply:
How about we blur the whole image only once, and then extract the rectangles and put them on top of the original image? Also, this discussion should (also) belong to the design doc. If it is not there, please make sure you describe the general idea in the design doc, and provide an analysis of the alternatives suggested here.

Reply:
I think the problem with this approach is that there has to be a different blur effect depending on each rectangle's size